Sources/Matrix.swift

```diff
@@ -0,0 +1,370 @@
+            return self.data[(row * columns) + column]
+        }
+        set {
+            precondition(validIndex(row: row, column: column ), MatrixError.invalid.localizedDescription)
+            self.data[(row * columns) + column] = newValue
+        }
+    }
+    
+    /**
+     :name:    validIndex
+     :description:    asserts that the (row,col) index is valid
+     - returns:    Bool
+     */
+    private func validIndex(row: Int, column: Int) -> Bool {
+        return row >= 0 && row < rows && column >= 0 && column < columns
+    }
+    
+    /**
+     :name:    column
+     :description:    gets the column using the index
+     - returns:    [Arithmetic]
+     */
+    func row(index: Int) -> [T] {
+        precondition(validIndex(row: index, column: 0), MatrixError.invalid.localizedDescription)
+        let row = Array(data[(index * columns)..<(index * columns) + columns])
+        return row
+    }
+    
+    /**
+     :name:    column
+     :description:    gets the column using the index
+     - returns:    [Arithmetic]
+     */
+    func column(index: Int) -> [T] {
+        assert(validIndex(row: 0, column: index), MatrixError.invalid.localizedDescription)
+        let column = (0..<rows).map { (currentRow) -> T in
+            let currentColumnIndex = currentRow * columns + index
+            return data[currentColumnIndex]
+        }
+        return column
+    }
+    
+    /**
+     :name:    transpose
+     :description:    mutates self to become the transpose matrix
+     - returns:    Void
+     */
+    mutating func transpose() {
+        var trans = Matrix.init(rows: self.columns, columns: self.rows, data: self.data)
+        for row in 0..<trans.rows {
+            for col in 0..<trans.columns {
+                trans[row,col] = self[col, row]
+            }
+        }
+        self = trans
+    }
+    
+    /**
+     :name:    transpose_copy
+     :description:    creates a copy of transpose matrix
+     - returns:    Matrix<T: Numeric>
+     */
+    func transpose_copy() -> Matrix<T> {
+        var trans = Matrix.init(rows: self.columns, columns: self.rows, data: self.data)
+        for row in 0..<trans.rows {
+            for col in 0..<trans.columns {
+                trans[row,col] = self[col, row]
+            }
+        }
+        return trans
+    }
+    
+}
+
+
+// MARK: Matrix Booleans
+extension Matrix {
+    
+    /**
+     :name:    isSquare
+     :description:    checks if the matrix has same number of rows and columns
+     - returns:    Bool
+     */
+    public var isSquare : Bool {
+        return self.rows == self.columns
+    }
+    
+    /**
+     :name:    isDiagonal
+     :description:    checks if the matrix is a square matrix and if it is, from i=0 to i = rowSize - 1 , if matrix[i,i] = 1.
+     - returns:    Bool
+     */
+    public var isDiagonal : Bool {
+        if !isSquare { return false }
```

Comment:
We generally then to use guard for instead of `if`.

```swift
guard !isSquare else {
  return false
}
```

Reply:
Sure, thats fine!